feat(ui): The status bar indicates that the panes are full screen and how many hidden panes are

takezyou opened the pull request:
Implementation of #168
I plan to display it in the tab bar instead of the status bar

phaazon commented:
Do you have a screenshot? Is there any ETA about the approval and merge of this PR?

takezyou commented:
@phaazon 
This will be the one I just completed.
<img width="1424" alt="Screen Shot 2021-09-05 at 23 37 39" src="https://user-images.githubusercontent.com/20274882/132130813-df311f66-2153-42b1-a2fd-a35f3a06827a.png">

imsnif commented on default-plugins/tab-bar/src/tab.rs:
```diff
@@ -40,12 +40,19 @@ pub fn non_active_tab(text: String, palette: Palette, separator: &str) -> LinePa
     let mut tab_text = text;
+    if is_fullscreen_active {
+        let current_panes = panes - 2;
+        let fullscreen_text = format!(" (Fullscreen +{} hidden panes)", current_panes);
```
Do you think it will be possible to have this in the status-bar instead? Maybe instead of the "Tip" line when we're in full-screen? I feel it's a better place for it.

TheLostLambda replied:
I'd second this! That was my first thought when seeing this :)

takezyou replied:
I think so too, so I'll try to fix that!

imsnif commented on zellij-server/src/screen.rs:
```diff
@@ -370,7 +370,9 @@ impl Screen {
             tab_data.push(TabInfo {
                 position: tab.position,
                 name: tab.name.clone(),
+                panes: tab.get_pane_num(),
```
When the full screen is active, all panes that are not full-screen are placed in tab.panes_to_hide... I think we can count them instead of doing this. What do you think?

takezyou replied:
I find it more convenient to use panes_to_hide, so I'll use that.

imsnif reviewed (changes requested):
Hey, I'm really happy you took this up again! :)

I left some comments. The main one is that I think this should be in the status bar instead of the tab bar. Specifically instead of the line that starts with "Tip:" whenever we're in full screen.

Looking forward to merge this feature! I know for me personally it will be super helpful. I'm in full screen a lot and sometimes forget that I am, and then open new panes and everything gets messed up :)

takezyou commented:
@imsnif 
I've made some corrections to the parts you reviewed.
However, I believe that the code and display of the UI part is not very good. Any advice would be appreciated!

https://user-images.githubusercontent.com/20274882/132722543-6a7566bf-6814-4232-836d-0bcc05741c58.mov

a-kenji commented:
Thanks @takezyou, this works already great.
There is still one more edge case that I saw while trying it out:

* start zellij
* open a pane
* go fullscreen
* open a pane

Now we are not in fullscreen anymore, but the status bar still shows
that we are.

takezyou commented:
@a-kenji 
Thank you!
I've made the corrections to the parts you pointed out.
Also fixed the second line information to be accurate.

takezyou commented:
@imsnif 
Thanks for the confirmation!
I'll check regarding the issue.

I'll try to change the visuals as you suggested.
I'll ask @a-kenji  when the fix is complete!

takezyou commented:
@a-kenji 
I have fixed some of the points raised and would appreciate your review.

a-kenji commented:
@takezyou 
This looks much better, thank you for all the work.
I would like to show this information also in the locked mode,
after the `-- Locked --` indicator, since some people spend more time in the
locked mode, than in other modes.

takezyou commented:
@a-kenji 
I fixed it.
Added display of interface locked when in full screen. What do you think?
<img width="1440" alt="Screen Shot 2021-09-16 at 0 34 06" src="https://user-images.githubusercontent.com/20274882/133463943-d8ce43b1-0fe9-4efb-ba02-f2edd392d4e9.png">

a-kenji commented:
@takezyou 
This is pretty good!
I would like a char to differentiate between the Locked and the 
Fullscreen, probably a `-` but then I am in general happy with the feature.

takezyou commented:
@a-kenji 
How about showing the original lock?
I feel like that would make it easier to distinguish between the two.

<img width="1440" alt="Screen Shot 2021-09-19 at 16 37 17" src="https://user-images.githubusercontent.com/20274882/133919417-47d4f578-d124-4039-829f-7fe0e898d12b.png">

a-kenji commented:
@takezyou 
I agree, that is much better!
Thank you for the great work.
If you want you can squash some commits, 
I will merge it later today.

takezyou commented:
@a-kenji 
Thank you!
I'll go ahead and squash some commits now.

takezyou commented:
@a-kenji 
All commits in one! Please confirm!

a-kenji reviewed (approved):
LGTM, I have tested this manually.
But the end to end tests fail all the time - do you maybe have an idea what this could be
@imsnif ?

Thanks for the work @takezyou

imsnif commented:
@a-kenji - from a brief glance it looks like there's a crash with a stacktrace in (all?) tests inside the container (this is distinct from the stacktrace of the failing test itself - you can see the tail end of it in the provided snapshot), eg.:
https://github.com/zellij-org/zellij/pull/450/checks?check_run_id=3654778354#step:12:2478

I would start by trying to replicate one test manually by SSHing into the container and doing all the steps that it does. You'll likely see the same stack trace and be able to understand what's causing it exactly.

I have a bit of a busy week, so not sure I'll have time to take a closer look soon unfortunately. If everyone is really stuck though I'll try to make time in one evening or the weekend.

imsnif commented:
Hum. Github doesn't seem to want to allow links to the build logs. Look for a line starting with "current_snapshot:" - this is a representation of what the test "sees on the screen", and in most cases there's the end of a stacktrace.

a-kenji commented:
@imsnif 
Thanks for the pointer!
This is what I see:
```
Timed out waiting for data on the SSH channel for test starts_with_one_terminal. Was waiting for step: Wait for app to load
cursor x: 2
cursor_y: 23
current_snapshot:
```
and then 
```
thread 'main' panicked at 'Timed out waiting for test', src/tests/e2e/remote_runner.rs:390:25
```
I will try to recreate this in a virtual machine, but I suspect this will be somewhat slow.

imsnif commented:
No need for a virtual machine (unless I'm missing something?)

On this branch, do:
1. cargo make build-e2e
2. Start the docker environment, ssh into the container (I think the user/password are test/test, but it should be specified in the docker-compose file)
3. Run zellij and do what the test does (eg. open a new pane)

a-kenji commented:
@imsnif 
Thanks!

> No need for a virtual machine (unless I'm missing something?)

I had trouble getting docker compose working on my machine, but it seems
the bug I had has been fixed now. So I can run it natively now.
If I understand correctly the e2e-test build the debug version, could that be the
reason for some of the timeouts? Especially the new tab one can take a long time in that
circumstance.

imsnif commented:
@a-kenji - as discussed offline, the e2e tests using the debug build was a very good find and I'm fixing it elsewhere. I don't think this is the issue here though. If you don't have time to take a look at this, I might be able to free some time up for this on the weekend. Let me know.

a-kenji commented:
@imsnif 
That is good to hear! Yes that is probably the plugin problem?
I might get time tomorrow to work on it - but I can not be sure yet.
If you get the time feel free to work on it. Maybe I will get to it and ping you!

a-kenji commented:
@imsnif 
I have gotten the tests running locally, they are super flaky.
I added the insta snapshot from the fullscreen change.

I think I found the issue for the flaky tests.
At least, if the test builds similar on gh to my local machine.
The release version runs fine. The musl release/debug builds sometimes
close panes on random for me. Can you check if that happens for you as well?

imsnif commented:
@a-kenji that's super weird. I just checked the musl release build and it runs just fine for me. What are you doing when it's closing panes for you?

I did a lot of work on the e2e tests today in the mirrored-sessions branch. It might be a pain to get them to work with this branch though (and I'm still not done - there's 1 test that often fails on the CI, but it specifically tests the mirrored sessions so you can safely ignore it).

a-kenji commented:
@imsnif 
I run `cargo make build-e2e`
And then run the musl build.
The release from our website is fine for me too.

> I did a lot of work on the e2e tests today in the mirrored-sessions branch. It might be a pain to get them to work with this branch though (and I'm still not done - there's 1 test that often fails on the CI, but it specifically tests the mirrored sessions so you can safely ignore it).

Should I push the changes I made?
They are fairly minor.

imsnif commented:
@a-kenji - what changes?

a-kenji commented:
> I added the insta snapshot from the fullscreen change.

And I changed from debug to release build, but there are still some issues.
But maybe that could help us get to the problem quicker.

imsnif commented:
If it still passes once you do (all except the one test there, the mirror test, that isn't currently passing) then sure

takezyou commented:
@imsnif 
Thank you for merging with us too!
If I can find another good isuue, I'll try it.
If you have any recommendations, let me know!

@a-kenji 
Thanks for all the advice!
I'm so glad this pull request got merged!

a-kenji commented:
@takezyou 
Thank you for keeping at it!
I am glad I could help.
